### PR TITLE
abci: export errors 

### PR DESCRIPTION
**Pull request by pysel**
---
Contributes: #1140 

#### Changes

- Exports errors in `abci`. More context: #1140.
- Small fix of deprecated code: [link](https://github.com/cometbft/cometbft/pull/1185/files#diff-5a2d69b7284d718471940b479f58d06ccbc918f6f38d0faa87afeeab463fa517R46)

### CODE REVIEW

**Comment by sergio-mena** [comment]
Thanks a lot for this @pysel !
One small comment: you probably want to replace "Closes #1140" by something like "Contributes to #1140" in the description.
We have a policy of not reviewing draft PRs until they are transitioned to ready-for-review PRs, unless the author explicitly requests it. Let us know if you want this reviewed before you transition it to ready-for-review.

**Comment by pysel** [comment]
> Thanks a lot for this @pysel !
> One small comment: you probably want to replace "Closes #1140" by something like "Contributes to #1140" in the description.
> We have a policy of not reviewing draft PRs until they are transitioned to ready-for-review PRs, unless the author explicitly requests it. Let us know if you want this reviewed before you transition it to ready-for-review.

sure thing, will change! 

this PR is still in draft, because I want to add more docs to newly introduced errors. will open for review as soon as I do that, thanks!

File: abci/client/errors.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	"github.com/cometbft/cometbft/abci/types"
+)
+
+// ErrUnknownAbciTransport is returned when trying to create a client with an invalid transport option
+type ErrUnknownAbciTransport struct {
+	Transport string
+}
+
+func (e ErrUnknownAbciTransport) Error() string {
+	return fmt.Sprintf("Unknown abci transport: %s", e.Transport)
```
Comment (thanethomson) [inline abci/client/errors.go:15]:
```suggestion
	return fmt.Sprintf("unknown abci transport: %s", e.Transport)
```

File: abci/client/errors.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+type ErrUnexpectedResponse struct {
+	Response types.Response
+	Comment  string
+}
+
+func (e ErrUnexpectedResponse) Error() string {
+	return fmt.Sprintf("Unexpected response %T. %s", e.Response.Value, e.Comment)
```
Comment (thanethomson) [inline abci/client/errors.go:24]:
```suggestion
	return fmt.Sprintf("unexpected response %T: %s", e.Response.Value, e.Reason)
```

File: abci/client/socket_client.go [hunk 3]
```diff
@@ -210,12 +210,12 @@ func (cli *socketClient) didRecvResponse(res *types.Response) error {
 	// Get the first ReqRes.
 	next := cli.reqSent.Front()
 	if next == nil {
-		return fmt.Errorf("unexpected response %T when no call was made", res.Value)
+		return ErrUnexpectedResponse{Response: *res, Comment: "No call was made"}
```
Comment (thanethomson) [inline abci/client/socket_client.go:213]:
```suggestion
		return ErrUnexpectedResponse{Response: *res, Comment: "no call was made"}
```

File: abci/client/socket_client.go [hunk 4]
```diff
@@ -210,12 +210,12 @@ func (cli *socketClient) didRecvResponse(res *types.Response) error {
...
 	}
 
 	reqres := next.Value.(*ReqRes)
 	if !resMatchesReq(reqres.Request, res) {
-		return fmt.Errorf("unexpected response %T to the request %T", res.Value, reqres.Request.Value)
+		return ErrUnexpectedResponse{Response: *res, Comment: fmt.Sprintf("Unexpected response to the request %T", reqres.Request.Value)}
```
Comment (thanethomson) [inline abci/client/socket_client.go:218]:
```suggestion
		return ErrUnexpectedResponse{Response: *res, Comment: fmt.Sprintf("unexpected response to the request %T", reqres.Request.Value)}
```

File: abci/server/errors.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	"github.com/cometbft/cometbft/abci/types"
+)
+
+// ErrUnknownServerType is returned when trying to create a server with invalid transport option.
+type ErrUnknownServerType struct {
+	ServerType string
+}
+
+func (e ErrUnknownServerType) Error() string {
+	return fmt.Sprintf("Unknown server type %s", e.ServerType)
```
Comment (thanethomson) [inline abci/server/errors.go:15]:
```suggestion
	return fmt.Sprintf("unknown server type %s", e.ServerType)
```

File: abci/server/errors.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+// ErrConnectionNotExists is returned when trying to access non-existent network connection
+type ErrConnectionNotExists struct {
+	ConnID int
+}
+
+func (e ErrConnectionNotExists) Error() string {
+	return fmt.Sprintf("Connection %d does not exist", e.ConnID)
```
Comment (thanethomson) [inline abci/server/errors.go:24]:
```suggestion
	return fmt.Sprintf("connection %d does not exist", e.ConnID)
```

File: abci/server/errors.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+type ErrUnknownClientRequest struct {
+	Req *types.Request
+}
+
+func (e ErrUnknownClientRequest) Error() string {
+	return fmt.Sprintf("Unknown request from client %T", e.Req)
```
Comment (thanethomson) [inline abci/server/errors.go:32]:
```suggestion
	return fmt.Sprintf("unknown request from client: %T", e.Req)
```

File: abci/server/errors.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+type ErrUnknownServerType struct {
+	ServerType string
+}
+
+func (e ErrUnknownServerType) Error() string {
+	return fmt.Sprintf("Unknown server type %s", e.ServerType)
+}
+
+// ErrConnectionNotExists is returned when trying to access non-existent network connection
+type ErrConnectionNotExists struct {
```
Comment (thanethomson) [inline abci/server/errors.go:19]: I'd recommend renaming this to `ErrConnectionDoesNotExist`.

File: abci/client/socket_client.go (hunk 4 above)
Comment (sergio-mena) [inline abci/client/socket_client.go:218]: Why passing the whole `Response`,  rather than just `Value`?

File: abci/client/errors.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+type ErrUnknownAbciTransport struct {
+	Transport string
+}
+
+func (e ErrUnknownAbciTransport) Error() string {
+	return fmt.Sprintf("Unknown abci transport: %s", e.Transport)
+}
+
+type ErrUnexpectedResponse struct {
+	Response types.Response
```
Comment (thanethomson) [inline abci/client/errors.go:19]: Why is this of type `types.Response` and, in `abci/server/errors.go`, `ErrUnknownClientRequest.Req` is of type `*types.Request`?

File: abci/server/errors.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	ConnID int
+}
+
+func (e ErrConnectionNotExists) Error() string {
+	return fmt.Sprintf("Connection %d does not exist", e.ConnID)
+}
+
+type ErrUnknownClientRequest struct {
+	Req *types.Request
```
Comment (thanethomson) [inline abci/server/errors.go:28]:
```suggestion
	Request *types.Request
```

To keep the same convention as `ErrUnexpectedResponse` in `abci/client/errors.go`.

File: abci/server/errors.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+// ErrConnectionNotExists is returned when trying to access non-existent network connection
+type ErrConnectionNotExists struct {
+	ConnID int
+}
+
+func (e ErrConnectionNotExists) Error() string {
+	return fmt.Sprintf("Connection %d does not exist", e.ConnID)
+}
+
+type ErrUnknownClientRequest struct {
```
Comment (thanethomson) [inline abci/server/errors.go:27]: `ErrUnexpectedRequest` perhaps?

File: abci/client/socket_client.go (hunk 4 above)
Comment (pysel) [reply abci/client/socket_client.go:218]: > 	Value isResponse_Value `protobuf_oneof:"value"`

the type of a Value is not exported by `types`. it is generated, so I personally do not think it is worth making it exported (at least if it is possible)

File: abci/client/errors.go (hunk 9 above)
Comment (pysel) [reply abci/client/errors.go:19]: no reason, going to remove the pointer

File: abci/server/errors.go (hunk 11 above)
Comment (pysel) [reply abci/server/errors.go:27]: wdyt of `ErrUnknownRequest`? I am changing this: `"unknown request from client %T"` with this error

Comment (pysel) [reply abci/server/errors.go:27]: @thanethomson

**Comment by pysel** [comment]
hey @thanethomson. review addressed!

**Review by thanethomson — APPROVED** [review]
LGTM, thanks @pysel!

**Pull request merged**